packages/ic-asset-certification/src/asset_router.rs

```diff
@@ -180,92 +181,29 @@ impl<'content> AssetRouter<'content> {
     /// given [HttpRequest](ic_http_certification::HttpRequest).
     pub fn serve_asset(
         &self,
+        data_certificate: &[u8],
         request: &HttpRequest,
-    ) -> AssetCertificationResult<(HttpResponse<'content>, HashTree, Vec<String>)> {
+    ) -> AssetCertificationResult<HttpResponse<'content>> {
         let preferred_encodings = self.get_preferred_encodings(request);
-        let req_path = request.get_path()?;
+        let request_url = request.get_path()?;
 
-        if let Some(CertifiedAssetResponse {
-            response,
-            tree_entry,
-        }) = self.get_encoded_asset(&preferred_encodings, &req_path)
-        {
-            let witness = self.tree.borrow().witness(tree_entry, &req_path)?;
-            let expr_path = tree_entry.path.to_expr_path();
-
-            return Ok((response.clone(), witness, expr_path));
-        }
-
-        if let Some(CertifiedAssetResponse {
-            response,
-            tree_entry,
-        }) = self.responses.get(&request_key(&req_path, None))
+        match self
+            .get_asset_for_request(&request_url, preferred_encodings)
+            .cloned()
         {
-            let witness = self.tree.borrow().witness(tree_entry, &req_path)?;
-            let expr_path = tree_entry.path.to_expr_path();
-
-            return Ok((response.clone(), witness, expr_path));
-        }
-
-        let mut url_scopes = req_path.split('/').collect::<Vec<_>>();
-        url_scopes.pop();
-
-        while !url_scopes.is_empty() {
-            let mut scope = url_scopes.join("/");
-            scope.push('/');
-
-            if let Some(CertifiedAssetResponse {
-                response,
-                tree_entry,
-            }) = self.get_encoded_fallback_asset(&preferred_encodings, &scope)
-            {
-                let witness = self.tree.borrow().witness(tree_entry, &req_path)?;
-                let expr_path = tree_entry.path.to_expr_path();
-
-                return Ok((response.clone(), witness, expr_path));
-            }
-
-            if let Some(CertifiedAssetResponse {
-                response,
+            Some(CertifiedAssetResponse {
+                mut response,
                 tree_entry,
-            }) = self.fallback_responses.get(&request_key(&scope, None))
-            {
-                let witness = self.tree.borrow().witness(tree_entry, &req_path)?;
+            }) => {
+                let witness = self.tree.borrow().witness(&tree_entry, &request_url)?;
                 let expr_path = tree_entry.path.to_expr_path();
 
-                return Ok((response.clone(), witness, expr_path));
-            }
-
-            scope.pop();
+                add_certificate_header(data_certificate, &mut response, &witness, &expr_path);
```

Comment:
If we use hard-coded v2-certification in the `AssetRouter` (via `add_certificate_header`), how will clients transition to v3?  Maybe certification version should be a property of `AssetRouter` set in the constructor?  🤔

Reply:
Let's discuss that next week, and maybe address in a separate PR.